hyperspy/model.py

```diff
@@ -924,475 +956,698 @@ def red_chisq(self):
-                                "signals is not equal to the variance shape "
-                                "of the signal")
-                    elif not isinstance(variance, numbers.Number):
-                        raise AttributeError(
-                            "Variance must be a number or a `Signal` instance "
-                            "but currently it is a %s" % type(variance))
-
-                weights = 1. / np.sqrt(variance)
-            else:
-                raise ValueError(
-                    'method must be "ls", "ml" or "custom" but %s given' %
-                    method)
-            args = (self.signal()[np.where(self.channel_switches)],
-                    weights)
 
-            # Least squares "dedicated" fitters
+            signal_shape = self.signal.data.shape
+            if isinstance(weights, np.ndarray):
+                if weights.shape != signal_shape:
+                    raise ValueError(
+                        "weights must have the same shape as the signal. "
+                        f"Got {weights.shape}, expected {signal_shape}"
+                    )
+            elif weights == "inverse_variance":
+                weights = self._convert_variance_to_weights()
+
+            args = (self.signal()[np.where(self.channel_switches)], weights)
+
             if fitter == "leastsq":
                 if bounded:
-                    # leastsq with bounds requires scipy >= 0.17
-                    if LooseVersion(
-                            scipy.__version__) < LooseVersion("0.17"):
-                        raise ImportError(
-                            "leastsq with bounds requires SciPy >= 0.17")
+                    self.set_boundaries(bounded=bounded)
 
-                    self.set_boundaries()
                     ls_b = self.free_parameters_boundaries
-                    ls_b = ([a if a is not None else -np.inf for a, b in ls_b],
-                            [b if b is not None else np.inf for a, b in ls_b])
-                    output = least_squares(self._errfunc, self.p0[:],
-                                           args=args, bounds=ls_b, **kwargs)
-                    self.p0 = output.x
-                    ysize = len(output.fun)
-                    cost = 2 * output.cost  # res.cost is half sum of squares
+                    ls_b = (
+                        [a if a is not None else -np.inf for a, b in ls_b],
+                        [b if b is not None else np.inf for a, b in ls_b],
+                    )
+
+                    if grad is True:
+
+                        def _wrap_jac(*args, **kwargs):
+                            # Jacobian function computes derivatives along
+                            # columns, so we need the transpose instead
+                            return self._jacobian(*args, **kwargs).T
+
+                        jacobian = _wrap_jac
+
+                    elif grad is False or grad is None:
```

Comment:
Then maybe make ``"2-point"`` the default and deprecate ``None`` and ``False``?

Reply:
Yep.